sound/soc/sof/sof-priv.h

```diff
@@ -118,6 +118,52 @@ struct snd_sof_platform_stream_params {
 	bool no_ipc_position;
 };
 
+struct snd_sof_dsp_ipc_ops {
+	/* doorbell */
+	irqreturn_t (*irq_handler)(int irq, void *context); /* optional */
+	irqreturn_t (*irq_thread)(int irq, void *context); /* optional */
+
+	/* ipc */
+	int (*send_msg)(struct snd_sof_dev *sof_dev,
+			struct snd_sof_ipc_msg *msg); /* mandatory */
```

Comment:
> The `msg` parameter for `ipc_msg_data()` / `send_msg()` is always exclusively `&sdev->ipc->msg` The RX (reply or notification) also uses the `sdev->ipc->msg`, I don't think that there is even a slight possibility to give 'custom' `struct snd_sof_ipc_msg` as we need to have serialization among messages. Would it make sense to drop the `msg` around this move?

that would need to be a separate change. We can do this before this PR, or after this PR. I don't want to add any functionality while splitting the existing code in separate ops.

Reply:
> We also need another layer of ops on top for core IPC level handling, or if I think about it we might not need them at all as the functions calling tx_message, set_get_data and friends have been moved to IPC specific ops, thus they can just happily call IPC specific functions?
> 
> Fwiw I have mostly implemented these for IPC3:
> 
> ```c
> struct ipc_ops {
>  	const struct ipc_tplg_ops *tplg;
>  	const struct ipc_pm_ops *pm;
>  	const struct ipc_pcm_ops *pcm;
> +
> +	int (*tx_msg)(struct snd_sof_dev *sdev, void *msg_data, size_t msg_bytes,
> +		      void *reply_data, size_t reply_bytes, bool no_pm);
> +	void (*rx_msg)(struct snd_sof_dev *sdev);
> +	int (*set_data)(struct snd_sof_dev *sdev, void *data, size_t data_bytes);
> +	int (*get_data)(struct snd_sof_dev *sdev, void *data, size_t data_bytes);
>  };
> ```
not following this comment at all. What PR are you referring to @ujfalusi ?

Reply:
@plbossart, I don't have a PR as my wip is based on several PRs in flight and have daily conflicts when rebasing on the updated PRs.

The send_msg() here is the platform IPC message sending, which can differ between IPC versions.
In top level SOF we have the user facing tx_message() variants for IPC3 and IPC4, they are calling this platform callback, but they do IPC specific lifting as well.
So, the question is: do we want ops for that level or we are calling the functions from already IPC dependent code.
I have a generic tx_msg() for ipc3/4 and set_get_data() for ipc3 implemented. We can have some shared code left in ipc.c, but most of the content there atm is realy ipc3 only.

Reply:
The intent of the ops I added was to cover the low-level combination of hardware specific + IPC type specific code.

The interrupt handling falls in this category, since we are using different registers for IPC between hardware generations.

I am not completely sure if there is merit is abstracting in sof_dsp_ipc_ops() any functionality that is NOT hardware specific. That should be handled in the higher level code of ipcv3/ipcv4.

So with that, I am not sure if the partitioning I chose is actually relevant. I don't think there's any issue with 
- irq_handler
- irq_thread
- fw_ready

I am not sure about 
- send_msg
- ipc_msg_data

I am even less sure about 
- ipc_dump. I don't know if we use hardware-specific resources here.
- set_stream_data_offset. This is a bit inconsistent because I've left get_mailbox_offset() and get_window_offset() out.

In other words, my main ask is to review what needs to be in these sof_dsp_ipc_ops structures, and what is handled at a higher level.

Reply:
`send_msg` iand `ipc_msg_data` s really IPC specific, the sending of a message is done differently for IPC3 compared to IPC4, just enough that it is cleaner to have different ops.
`ipc_dump` is more like platform than IPC related by the implementations.
`set_stream_data_offset`, is probably again fall to platform along with the get_mailbox_offset() and window_offset()?

About the `snd_sof_ipc_set_get_comp_data()`, `sof_ipc_tx_message()` and `sof_ipc_tx_message_no_pm()`, we don't want them to be a generic interface, but rename them like s/ipc/ipc3 and have ipc4 variants (we already have `ipc4_tx_message()`) when needed, right?
Related things, like `tx_wait_done()` and `ipc_log_header()` are IPC specific and not that reusable, unless we move the wait down to platform level, within `snd_sof_dsp_send_msg()` callback.

Reply:
Sorry @ujfalusi I've lost me. I am not that familiar with the IPC and I don't know what is remark/question/suggestion/recommendation.

It'd be more productive for me if you suggested a code diff or patch on what you think would make sense.

Reply:
@plbossart, OK, I'll have a draft tomorrow. I have the IPC3 now working, but I'm not happy with the code duplication between IPC versions.